src/ty/variant.rs

```diff
@@ -0,0 +1,237 @@
+			variants: registry.map_into_compact(self.variants),
+		}
+	}
+}
+
+impl TypeVariant {
+	#[cfg_attr(feature = "cargo-clippy", allow(clippy::new_ret_no_self))]
+	pub fn new(name: &'static str, namespace: Namespace) -> TypeVariantBuilder {
+		TypeVariantBuilder {
+			path: Path::new(name, namespace),
+		}
+	}
+}
+
+pub struct TypeVariantBuilder {
+	path: PathBuilder,
+}
+
+impl TypeVariantBuilder {
+	pub fn type_params<I>(self, type_params: I) -> Self
+	where
+		I: IntoIterator<Item = MetaType>,
+	{
+		let mut this = self;
+		this.path.type_params(type_params);
+		this
+	}
+
+	pub fn variants<F>(self, variants: Variants<F>) -> TypeVariant {
+		TypeVariant {
+			path: self.path.done(),
+			variants: variants.variants(),
+		}
+	}
+}
+
+/// A struct enum variant with either named (struct) or unnamed (tuple struct)
+/// fields.
+///
+/// # Example
+///
+/// ```
+/// enum Operation {
+///     Zero,
+/// //  ^^^^ this is a unit struct enum variant
+///     Add(i32, i32),
+/// //  ^^^^^^^^^^^^^ this is a tuple-struct enum variant
+///     Minus { source: i32 }
+/// //  ^^^^^^^^^^^^^^^^^^^^^ this is a struct enum variant
+/// }
+/// ```
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug, Serialize)]
+#[serde(bound = "F::TypeId: Serialize")]
+pub struct Variant<F: Form = MetaForm> {
+	/// The name of the struct variant.
+	name: F::String,
+	/// The fields of the struct variant.
+	#[serde(skip_serializing_if = "Vec::is_empty")]
+	fields: Vec<Field<F>>,
+	/// The discriminant of the variant.
+	///
+	/// # Note
+	///
+	/// Even though setting the discriminant is optional
+	/// every C-like enum variant has a discriminant specified
+	/// upon compile-time.
+	#[serde(skip_serializing_if = "Option::is_none")]
+	discriminant: Option<u64>,
+}
+
+impl IntoCompact for Variant {
+	type Output = Variant<CompactForm>;
+
+	fn into_compact(self, registry: &mut Registry) -> Self::Output {
+		Variant {
+			name: registry.register_string(self.name),
+			fields: registry.map_into_compact(self.fields),
+			discriminant: self.discriminant,
+		}
+	}
+}
+
+impl Variant {
+	/// Creates a new variant with the given fields.
+	pub fn with_fields<F>(name: <MetaForm as Form>::String, fields: F) -> Self
+	where
+		F: IntoIterator<Item = Field>,
+	{
+		Self {
+			name,
+			fields: fields.into_iter().collect(),
+			discriminant: None,
+		}
+	}
+
+	/// Creates a new variant with the given discriminant.
+	pub fn with_discriminant(name: <MetaForm as Form>::String, discriminant: u64) -> Self {
+		Self {
+			name,
+			fields: Vec::new(),
+			discriminant: Some(discriminant),
+		}
+	}
+}
+
+/// Build a type with no variants.
+pub enum NoVariants {}
+/// Build a type where *any* variants consist of fields.
+pub enum VariantFields {}
+/// Build a type where *all* variants have no fields and a discriminant (e.g. a
+/// Clike enum)
+pub enum Discriminant {}
+
+pub struct Variants<T = NoVariants> {
+	variants: Vec<Variant>,
+	marker: PhantomData<fn() -> T>,
+}
```

Comment:
So this is basically a builder. But I guess the name is fine.

Reply:
Again you could split `Variants` into `Variants` and `VariantsBuilder` to get rid of the default type argument.

Reply:
> So this is basically a builder. But I guess the name is fine.
Yeah this is just to provide the DSL syntax: e.g. `Variants::with_fields()`